<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>CC_GWAS — CC_GWAS • CaseControlAF</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="CC_GWAS — CC_GWAS"><meta property="og:description" content="This is a function to run a case-case GWAS using the framework introduced in the ReACt package. The method will first filter out any SNPs with NA in the case or control AF. It will then remove any SNPs that are not present in both datasets. Finally, it will remove duplicate SNPs, by defualt keeping the first record. Then the CC-GWAS will be run."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CaseControlAF</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>CC_GWAS</h1>
    
    <div class="hidden name"><code>CC_GWAS.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This is a function to run a case-case GWAS using the framework introduced in the ReACt package. The method will first filter out any SNPs with NA in the case or control AF. It will then remove any SNPs that are not present in both datasets. Finally, it will remove duplicate SNPs, by defualt keeping the first record. Then the CC-GWAS will be run.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">CC_GWAS</span><span class="op">(</span></span>
<span>  <span class="va">trait1_data</span>,</span>
<span>  <span class="va">trait2_data</span>,</span>
<span>  AF_case_colname <span class="op">=</span> <span class="st">"AF_case"</span>,</span>
<span>  AF_control_colname <span class="op">=</span> <span class="st">"AF_control"</span>,</span>
<span>  <span class="va">trait1_N_case</span>,</span>
<span>  <span class="va">trait1_N_control</span>,</span>
<span>  <span class="va">trait2_N_case</span>,</span>
<span>  <span class="va">trait2_N_control</span>,</span>
<span>  chromosome_colname <span class="op">=</span> <span class="st">"chr"</span>,</span>
<span>  position_colname <span class="op">=</span> <span class="st">"pos"</span>,</span>
<span>  id_colname <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  A1_colname <span class="op">=</span> <span class="st">"A1"</span>,</span>
<span>  A2_colname <span class="op">=</span> <span class="st">"A2"</span>,</span>
<span>  rsid_colname <span class="op">=</span> <span class="st">"RSID"</span>,</span>
<span>  OR_colname <span class="op">=</span> <span class="st">"OR"</span>,</span>
<span>  SE_colname <span class="op">=</span> <span class="st">"SE"</span>,</span>
<span>  N_caseInCase <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  N_controlInControl <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  N_caseInControl <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  Zthreshold <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>trait1_data</dt>
<dd><p>a dataframe containing the chromosome, position, case AF, and control AF for each SNP for trait 1</p></dd>


<dt>trait2_data</dt>
<dd><p>a dataframe containing the chromosome, position, case AF, and control AF for each SNP for trait 2</p></dd>


<dt>AF_case_colname</dt>
<dd><p>a string with the name for the column containing the case AF in both trait 1 and trait 2 datasets; default: "AF_case"</p></dd>


<dt>trait1_N_case</dt>
<dd><p>the number of cases in trait 1 dataset</p></dd>


<dt>trait1_N_control</dt>
<dd><p>the number of controls in trait 1 dataset</p></dd>


<dt>trait2_N_case</dt>
<dd><p>the number of cases in trait 2 dataset</p></dd>


<dt>trait2_N_control</dt>
<dd><p>the number of controls in trait 2 dataset</p></dd>


<dt>chromosome_colname</dt>
<dd><p>a string with the name for the column containing the SNP chromosome number in both trait 1 and trait 2 datasets; default: "chr"</p></dd>


<dt>position_colname</dt>
<dd><p>a string with the name for the column containing the SNP position in both trait 1 and trait 2 datasets; default: "pos"</p></dd>


<dt>id_colname</dt>
<dd><p>a string containing the id for each SNP in both trait 1 and trait 2 datasets; default: NA. If NA, then an ID column will be added and used for data filtering</p></dd>


<dt>OR_colname</dt>
<dd><p>a string with the name for the column with the OR from the GWAS summary statistics for each trait</p></dd>


<dt>SE_colname</dt>
<dd><p>a string with the name for the column with the SE from the GWAS summary statistics for each trait</p></dd>


<dt>N_caseInCase</dt>
<dd><p>a 2x2 matrix where the upper left cell (a) is the number of cases in study 1 and the lower right cell (d) is the number of cases in study 2. Cells b and c are the number of overlapping cases; Default = NA - if left as NA algorithm will assume no sample overlap</p></dd>


<dt>N_controlInControl</dt>
<dd><p>a 2x2 matrix where the upper left cell (a) is the number of controls in study 1 and the lower right cell (d) is the number of controls in study 2. Cells b and c are the number of overlapping controls; Default = NA - if left as NA algorithm will assume no sample overlap</p></dd>


<dt>N_caseInControl</dt>
<dd><p>a 2x2 matrix of the number of cases in the control sample. The upper right cell (b) represents the number of cases from trait 1 that are present as controls for trait 2. The lower left cell (c) is the number of cases from trait 2 that are included as controls for trait 1. Cells a and d should be 0; Default = NA - if left as NA algorithm will assume no sample overlap</p></dd>


<dt>Zthreshold</dt>
<dd><p>value for the z-statistic threshold. If Zthreshold = 0 then will not perform correction for sample overlap (In order to run correction for sample overlap, OR and SE in each trait's data is required). Yang et al suggest using a threshold of z=1.0; however, this will overlap user input sample sizes. If exact sample sizes and overlap is known, this parameter should be left at 0.</p></dd>


<dt>AF_control_colnamea</dt>
<dd><p>a string with the name for the column containing the control AF in both trait 1 and trait 2 datasets; default: "AF_control"</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>returns a dataframe with the original data (where each dataset's original columns with have .trait1 or .trait2 appended to its column name) and three additional columns for the CC-GWAS results: OR, SE, p
equal to the number of variants</p>
    </div>
    <div id="note">
    <h2>Note</h2>
    <p>trait1_data and trait2_data need to have the same column names for chromosome, position, OR, SE, and case and control AFs.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by The package maintainer.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

